### DISCUSSION TITLE
Minimizing a quadratic form with Hermitian matrix

### DISCUSSION THREAD

**Discussion by alhernandezanton** [cvxpy/cvxpy#2222] [Q&A]
Hi, I need want to solve the following problem. I want to fit a matrix generated by a model (`rho`) to a Hermitian matrix of complex numbers (`rho_exp`) which has been measured experimentally. This array is the result of averaging over many measurements and thus has an associated covariance matrix (`cov`). The way I want to do this fit is by minimizing the modulus of the difference array d = (rho_exp - rho).flatten(), weighted by the uncertainties of the measurements. One way to define this is with the quadratic form Q = d*(cov^-1)d, where * stands for complex conjugation. This function takes real values and must be minimized. The model `rho` must be subject to a certain set of constraints. The code I use for such purpose is the following:

```
rho = cp.Variable((d, )*2, hermitian = True) # Define variable, which is a Hermitian matrix
rho_flat = cp.reshape(rho, (d**2,), 'C') # Reshape into a 1D object to be used as a vector

rho_exp_flat = rho_exp.flatten()
d = rho_exp_flat - rho_
W = np.linalg.inv(cov) # Weight matrix, defined as the inverse of the covariance matrix

Q = cp.quad_form(delta, W, assume_PSD = True) # Define the quadratic form 
objective = cp.Minimize(Q) # We want it to be minimized
constraints = [ # I need the variable rho to meet these conditions
    rho >> 0,                # PSD
    cp.trace(rho) == 1  # Unit trace
]

problem = cp.Problem(objective, constraints)
problem.solve()
```

When I do this, I get an assertion error, because the solver goes through a reduction called 'complex2real', which checks if there are no imaginary values in the W matrix (I think). This quadratic function takes positive real values and thus there should be no problem to optimize it. Does anyone know how to solve this issue?

Thanks in advance

**Comment by SteveDiamond** [comment]
Can you share a full script that replicates the error? It's not clear from this code why it wouldn't work. My instinct is it's related to using ``assume_PSD=True`` but I'm not sure.

**Comment by seventeenmagpie** [comment]
I got the same issue, solving a similar looking problem with the quadratic form `x^H @ Q @ x` (where Q is itself a Hermitian matrix). I was able to avoid the assert by formulating the objective directly as `obj = x.conj().T @ Q @ x` rather than using `quad_form`.
